src/set.rs

```diff
@@ -107,5 +146,385 @@ where
+        }
+        // No key kicks another one out.
+        for i in 0..TEST_BATCH_SMALL {
+            assert_eq!(i, set.insert(i).unwrap());
+        }
+    }
+
+    #[test]
+    fn simple_remove() {
+        let set = ConSet::new();
+        assert!(set.remove(&42).is_none());
+        assert!(set.insert(42).is_none());
+        assert_eq!(42, set.get(&42).unwrap());
+        assert_eq!(42, set.remove(&42).unwrap());
+        assert!(set.get(&42).is_none());
+        assert!(set.is_empty());
+        assert!(set.remove(&42).is_none());
+        assert!(set.is_empty());
+    }
+
+    fn remove_many_inner<H: BuildHasher>(mut set: ConSet<usize, H>, len: usize) {
+        for i in 0..len {
+            assert!(set.insert(i).is_none());
+        }
+        for i in 0..len {
+            assert_eq!(i, set.get(&i).unwrap());
+            assert_eq!(i, set.remove(&i).unwrap());
+            assert!(set.get(&i).is_none());
+            set.raw.assert_pruned();
+        }
+
+        assert!(set.is_empty());
+    }
+
+    #[test]
+    fn remove_many() {
+        remove_many_inner(ConSet::new(), TEST_BATCH);
+    }
+
+    #[test]
+    fn remove_many_collision() {
+        remove_many_inner(ConSet::with_hasher(NoHasher), TEST_BATCH_SMALL);
+    }
+
+    #[test]
+    fn collision_remove_one_left() {
+        let mut set = ConSet::with_hasher(NoHasher);
+        set.insert(1);
+        set.insert(2);
+
+        set.raw.assert_pruned();
+
+        assert!(set.remove(&2).is_some());
+        set.raw.assert_pruned();
+
+        assert!(set.remove(&1).is_some());
+
+        set.raw.assert_pruned();
+        assert!(set.is_empty());
+    }
+
+    #[test]
+    fn collision_remove_one_left_with_str() {
+        let mut set = ConSet::with_hasher(NoHasher);
+        set.insert("hello");
+        set.insert("world");
+
+        set.raw.assert_pruned();
+
+        assert!(set.remove("world").is_some());
+        set.raw.assert_pruned();
+
+        assert!(set.remove("hello").is_some());
+
+        set.raw.assert_pruned();
+        assert!(set.is_empty());
+    }
+
+    #[test]
+    fn remove_par() {
+        let mut set = ConSet::new();
+        for i in 0..TEST_THREADS * TEST_BATCH {
+            set.insert(i);
+        }
+
+        thread::scope(|s| {
+            for t in 0..TEST_THREADS {
+                let set = &set;
+                s.spawn(move |_| {
+                    for i in 0..TEST_BATCH {
+                        let num = t * TEST_BATCH + i;
+                        let val = set.remove(&num).unwrap();
+                        assert_eq!(num, val);
+                        assert_eq!(num, val);
+                    }
+                });
+            }
+        })
+        .unwrap();
+
+        set.raw.assert_pruned();
+        assert!(set.is_empty());
+    }
+
+    fn iter_test_inner<S: BuildHasher>(set: ConSet<(usize, usize), S>) {
+        for i in 0..TEST_BATCH_SMALL {
+            assert!(set.insert((i, i)).is_none());
```

Comment:
Is here (an in the few following tests) a particular reason to insert tuples, instead of just `i`, or is it just copying and not adjusting the map-based test?

Reply:
Nope; I tried just to stick with the _map_ tests but indeed is a bit stretched. I'll simplify.